batch/test/test_copy_files.py

```diff
@@ -0,0 +1,547 @@
+import os
+import unittest
+import tempfile
+import subprocess as sp
+import uuid
+import random
+import string
+from shlex import quote as shq
+import google.oauth2.service_account
+from hailtop.auth import get_userinfo
+from hailtop.utils.os import _glob
+
+from batch.google_storage import GCS
+
+key_file = '/gsa-key/key.json'
+project = os.environ['PROJECT']
+user = get_userinfo()
+tmp_bucket = f'gs://{user["bucket_name"]}/test_copy_files'
+
+credentials = google.oauth2.service_account.Credentials.from_service_account_file(key_file)
+gcs_client = GCS(None, project=project, credentials=credentials)
+
+
+class RemoteTemporaryDirectory:
+    def __init__(self):
+        token = uuid.uuid4().hex[:6]
+        self.name = tmp_bucket + f'/{token}'
+
+    def __enter__(self):
+        return self.name
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        remove_remote_dir(self.name)
+
+
+def upload_files(src, dest):
+    os.system(f'gsutil -m -q cp -r {src} {dest}')
+
+
+def move_files(src, dest):
+    os.system(f'gsutil -m -q mv -r {src} {dest}')
+
+
+def remove_remote_dir(path):
+    os.system(f'gsutil -m -q rm -r {path}')
+
+
+def touch_file(path, data=None):
+    dir = os.path.dirname(path)
+    if not os.path.isdir(dir):
+        os.makedirs(dir, exist_ok=True)
+
+    with open(path, 'w') as file:
+        if data:
+            file.write(data)
+
+
+def cp_batch(src, dest, parallelism=1, min_partition_size='1Gi',
+             max_upload_partitions=32, max_download_partitions=32):
+    cmd = f'''
+python3 -u -m batch.worker.copy_files \\
+  --key-file {key_file} \\
+  --project {project} \\
+  --parallelism {parallelism} \\
+  --min-partition-size {min_partition_size} \\
+  --max-upload-partitions {max_upload_partitions} \\
+  --max-download-partitions {max_download_partitions} \\
+  -f {shq(src)} {shq(dest)} 2>&1
+'''
+    result = sp.run(cmd, shell=True, stdout=sp.PIPE, stderr=sp.PIPE)
+    return str(result.stdout), result.returncode
```

Comment:
Yeah, you're already using subprocess here, I'd avoid the shell entirely so you don't have any quoting nonsense:

```
result = sp.run(['python3', '-u', '-m', 'batch.worker.copy_files',
                 '--key-file', key_file,
....
                 '-f', src, dest], stdout=sp.PIPE, stderr=sp.STDOUT)
```